File: java/0102_levelOrder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.Arrays;
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        traverse(0,root,res);
        return res;
    }
    public void traverse(int depth,TreeNode root,List<List<Integer>> res)
    {
        if(root==null)
            return;
        
        if(res.size()>depth)
            res.get(depth).add(root.val);
        else
        {
            var t = new ArrayList<Integer>();
            t.add(root.val);
            res.add(t);
        }
        traverse(depth+1,root.left,res);
        traverse(depth+1,root.right,res);
    }
}

File: cpp/0102_levelOrder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// 用一个depth变量记录当前的深度
// 前序遍历二叉树时按照深度添加到结果数组里
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        traverse(0,root,res);
        return res;
    }
    void traverse(int depth, TreeNode* root, vector<vector<int>> &res)
    {
        if(root==nullptr)
            return;
        
        if(depth < res.size())
            res[depth].push_back(root->val);
        else
            res.push_back({root->val});
        traverse(depth+1,root->left,res);
        traverse(depth+1,root->right,res);
    }
};
